File: src/main/java/org/cobbzilla/util/http/PooledHttpClientFactory.java
package org.cobbzilla.util.http;

import org.apache.http.HttpHost;
import org.apache.http.conn.routing.HttpRoute;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
import org.cobbzilla.util.reflect.ObjectFactory;

import java.util.Map;

public class PooledHttpClientFactory implements ObjectFactory<CloseableHttpClient> {

    private String host;
    private int maxConnections;

    @java.beans.ConstructorProperties({"host", "maxConnections"})
    public PooledHttpClientFactory(String host, int maxConnections) {
        this.host = host;
        this.maxConnections = maxConnections;
    }

    @Override public CloseableHttpClient create() {
        final PoolingHttpClientConnectionManager cm = new PoolingHttpClientConnectionManager();
        cm.setMaxTotal(maxConnections);
        cm.setMaxPerRoute(new HttpRoute(new HttpHost(host)), maxConnections);
        return HttpClients.custom()
                .setConnectionManager(cm)
                .setConnectionManagerShared(true)
                .build();
    }

    @Override public CloseableHttpClient create(Map<String, Object> ctx) { return create(); }

    public String getHost() {
        return this.host;
    }

    public int getMaxConnections() {
        return this.maxConnections;
    }
}

File: src/main/java/org/cobbzilla/util/http/PooledHttpClientFactory.kt
package org.cobbzilla.util.http

import org.apache.http.HttpHost
import org.apache.http.conn.routing.HttpRoute
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.impl.client.HttpClients
import org.apache.http.impl.conn.PoolingHttpClientConnectionManager
import org.cobbzilla.util.reflect.ObjectFactory

class PooledHttpClientFactory @java.beans.ConstructorProperties("host", "maxConnections")
constructor(val host: String, val maxConnections: Int) : ObjectFactory<CloseableHttpClient> {

    override fun create(): CloseableHttpClient {
        val cm = PoolingHttpClientConnectionManager()
        cm.maxTotal = maxConnections
        cm.setMaxPerRoute(HttpRoute(HttpHost(host)), maxConnections)
        return HttpClients.custom()
                .setConnectionManager(cm)
                .setConnectionManagerShared(true)
                .build()
    }

    override fun create(ctx: Map<String, Any>): CloseableHttpClient {
        return create()
    }
}
